drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -967,6 +973,136 @@ static int siwx91x_dev_init(const struct device *dev)
 	return 0;
 }
 
+static int siwx91x_convert_z_sl_twt_req_type(enum wifi_twt_setup_cmd z_req_cmd)
+{
+	switch (z_req_cmd) {
+	case WIFI_TWT_SETUP_CMD_REQUEST:
+		return REQUEST_TWT;
+	case WIFI_TWT_SETUP_CMD_SUGGEST:
+		return SUGGEST_TWT;
+	case WIFI_TWT_SETUP_CMD_DEMAND:
+		return DEMAND_TWT;
+	default:
+		return -EINVAL;
+	}
+}
+
+static int siwx91x_set_twt_setup(struct wifi_twt_params *params)
+{
+	sl_status_t status;
+	int twt_req_type = siwx91x_convert_z_sl_twt_req_type(params->setup_cmd);
+
+	if (twt_req_type < 0) {
+		params->fail_reason = WIFI_TWT_FAIL_CMD_EXEC_FAIL;
+		return -EINVAL;
+	}
+
+	if (!params->setup.twt_info_disable) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	if (params->setup.responder) {
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	/* implicit -> won't do renegotiation
+	 * explicit -> must do renegotiation for each session
+	 */
+	if (!params->setup.implicit) {
+		/* explicit twt is not supported */
+		params->fail_reason = WIFI_TWT_FAIL_OPERATION_NOT_SUPPORTED;
+		return -ENOTSUP;
+	}
+
+	sl_wifi_twt_request_t twt_req = {
+		.wake_duration_unit = 0,
+		.wake_int_mantissa = params->setup.twt_mantissa,
+		.un_announced_twt = !params->setup.announce,
+		.wake_duration = params->setup.twt_wake_interval,
+		.triggered_twt = params->setup.trigger,
+		.wake_int_exp = params->setup.twt_exponent,
+		.implicit_twt = 1,
+		.twt_flow_id = params->flow_id,
+		.twt_enable = 1,
+		.req_type = twt_req_type,
+	};
+
+	if (params->setup.twt_wake_interval > 255 * 256) {
+		twt_req.wake_duration_unit = 1;
+		twt_req.wake_duration = params->setup.twt_wake_interval / 256;
+	} else {
+		twt_req.wake_duration_unit = 0;
+		twt_req.wake_duration = params->setup.twt_wake_interval / 1024;
+	}
+
+	status = sl_wifi_enable_target_wake_time(&twt_req);
+	if (status != SL_STATUS_OK) {
+		params->fail_reason = WIFI_TWT_FAIL_CMD_EXEC_FAIL;
+		params->resp_status = WIFI_TWT_RESP_NOT_RECEIVED;
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+static int siwx91x_set_twt_teardown(struct wifi_twt_params *params)
+{
+	sl_status_t status = SL_STATUS_OK;
+	sl_wifi_twt_request_t twt_req;
```

Comment:
You don't need to initialize status, but you need to initialize `twt_req` (otherwise you are going to pass random values to `sl_wifi_disable_target_wake_time()`):
```
sl_wifi_twt_request_t twt_req = { };
```